### PR TITLE
always pass filterQuality specified in the Image widget to canvas

### LINKED ISSUE #73714: Provide higher quality image filter than `FilterQuality.high` (like Chrome's)
## Description

CanvasKit doesn't seem to smooth out low-res images in a way HTML renderer does. Normally, it would be best to provide resolution-aware assets (`images/2.0x/`, `images/3.0x/`), but in this case we have no control over the content (we get the HTML output from CMS API, parse it and render as Flutter widgets - including `` tags with CDN image links).

When rendered via HTML renderer, the small images are smoothed out as generally expected, but CanvasKit leaves the images very jaggy and rough. `filterQuality: FilterQuality.high` does not change anything, as it mostly helps with downsizing large resolution images.

## Screenshot
image

<details>
<summary>code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(ImageTestApp());
}

class ImageTestApp extends StatelessWidget {
  final isCanvasKit = const bool.fromEnvironment('FLUTTER_WEB_USE_SKIA', defaultValue: false);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        body: SingleChildScrollView(
          child: Container(
            padding: const EdgeInsets.all(8.0),
            width: double.infinity,
            child: Column(
              crossAxisAlignment: isCanvasKit ? CrossAxisAlignment.start : CrossAxisAlignment.end,
              children: <Widget>[
                Text(isCanvasKit ? 'CanvasKit' : 'HTML'),
                SizedBox(height: 20),
                Image(
                  image: AssetImage('images/sample2.png'),
                ),
                Image(
                  image: AssetImage('images/sample.png'),
                ),
                Image(
                  image: AssetImage('images/sample3.png'),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}
```
</details>

## Repository
[Repository w/ local image assets](https://github.com/kubaprzetakiewicz/canvaskit_jagged_images)

## Environment
```
➜ flutter doctor -v
[✓] Flutter (Channel master, 1.26.0-2.0.pre.276, on macOS 11.0.1 20B50 darwin-x64, locale en-GB)
    • Flutter version 1.26.0-2.0.pre.276 at /Users/kubcio/.flutter
    • Framework revision 08771fa783 (64 minutes ago), 2021-01-11 11:19:03 -0500
    • Engine revision 87004b05ab
    • Dart version 2.12.0 (build 2.12.0-204.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/kubcio/Library/Android/sdk
    • Platform android-30, build-tools 30.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    • CocoaPods version 1.10.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] IntelliJ IDEA Ultimate Edition (version 2020.3)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    • Flutter plugin version 52.1.5
    • Dart plugin version 203.6912

[✓] VS Code (version 1.52.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.18.1

[✓] Connected device (1 available)
    • Chrome (web) • chrome • web-javascript • Google Chrome 87.0.4280.88

• No issues found!
```

### PR DESCRIPTION
**Pull request by yjbanov**
Always pass `filterQuality` specified in the Image widget to canvas.

We previously assumed that `filterQuality` only matters when we draw the image to a different size. However, parent transforms and physical pixel misalignment frequently causes degradation in quality.

Fixes https://github.com/flutter/flutter/issues/73714

### CODE REVIEW

**Comment by skia-gold** [comment]
Gold has detected about 9 untriaged digest(s) on patchset 1.
View them at https://flutter-gold.skia.org/cl/github/74854

**Comment by skia-gold** [comment]
Gold has detected about 6 untriaged digest(s) on patchset 2.
View them at https://flutter-gold.skia.org/cl/github/74854

**Comment by flar** [comment]
Have you run some common benchmarks like gallery transitions? This can potentially affect performance, but I hope not as it is the right thing to do.

File: packages/flutter/lib/src/widgets/image.dart [hunk 1]
```diff
@@ -930,11 +917,22 @@ class Image extends StatefulWidget {
   /// If non-null, this color is blended with each image pixel using [colorBlendMode].
   final Color? color;
 
-  /// Used to set the [FilterQuality] of the image.
+  /// The rendering quality of the image.
+  ///
+  /// If the image is of a high quality and its pixels are perfectly aligned
+  /// with the physical screen pixels, extra quality enhancement may not be
+  /// necessary. If so, then [FilterQuality.none] would be the most efficient.
```
Comment (flar) [inline packages/flutter/lib/src/widgets/image.dart:924]: Check with Skia, but I believe that they might already detect this case and optimize it.

I'm not sure if the same could be said of the html web-renderer, though.

**Review by Hixie — APPROVED** [review]
Code LGTM but as @flar said please keep a close eye on perf numbers when landing this.

**Review by dnfield — APPROVED** [review]

**Comment by flar** [comment]
An alternate solution if we see a perf issue:

Change the property default to null. Then the quality would be:
`quality = property ?? (boxfit changed size ? low : null);`

**Comment by yjbanov** [comment]
Here are some benchmark results (averaged over 5 runs) from my Moto E5 (a ~$100 phone from 2018):

```
                                               BEFORE     AFTER
average_frame_rasterizer_time_millis             9.41      9.35
90th_percentile_frame_rasterizer_time_millis    15.75     15.99
99th_percentile_frame_rasterizer_time_millis    42.06     42.82
worst_frame_rasterizer_time_millis             118.10    113.19
missed_frame_rasterizer_budget_count            85.20     88.80
frame_count                                    807.00    810.00
frame_rasterizer_count                         807.60    811.00
```

[Full test results](https://docs.google.com/spreadsheets/d/1K_ublvQLWqO1uR76Zr_HkJ6Ri240H9p7K6APmjOH6CM/edit#gid=0).

I also manually benchmarked the "Shrine" sample app in the gallery, which is quite image-heavy. The difference between status quo and my change seems to be lost in the noise. However, when comparing rendering all images in `FilterQuality.none` vs `FilterQuality.high` I do see a slight bump in the rasterizer thread (from ~10.5ms to ~11ms) according to the performance overlay.

If this looks good, I'm ready to submit the change. Otherwise, open to discussing alternatives, such as allowing `null` filter quality that uses the current strategy suggested by @flar.

**Comment by Hixie** [comment]
Sounds good. As above, just keep an eye on the benchmarks when you land. :-)

**Comment by flar** [comment]
Yes, that all looks great! Tagging @liyuqian as I'm sure he'd be interested and so he's aware of what's coming down the PR pipeline...

**Review by flar — APPROVED** [review]
LGTM

**Review by liyuqian — APPROVED** [review]
Thanks! I'll keep an eye on this if I see some performance regressions in the near future.

**Comment by skia-gold** [comment]
Gold has detected about 12 untriaged digest(s) on patchset 3.
View them at https://flutter-gold.skia.org/cl/github/74854

**Comment by ranulfo** [comment]
I have a very similar problem reported here #129721

**Pull request merged**